test/BulletDynamics/test_deterministic_save_and_restore.cpp

```diff
@@ -0,0 +1,221 @@
+	{
+		btTransform groundTransform;
+		groundTransform.setIdentity();
+		groundTransform.setOrigin(btVector3(0, -50, 0));
+		btScalar mass(0.);
+		bool isDynamic = (mass != 0.f);
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			groundShape->calculateLocalInertia(mass, localInertia);
+
+		#ifdef USE_MOTIONSTATE
+			btDefaultMotionState* myMotionState = new btDefaultMotionState(groundTransform);
+
+			btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, groundShape, localInertia);
+
+			btRigidBody* body = new btRigidBody(cInfo);
+		#else
+			btRigidBody* body = new btRigidBody(mass, 0, groundShape, localInertia);
+			body->setWorldTransform(groundTransform);
+		#endif
+
+		body->setUserIndex(-1);
+		initial_world->addRigidBody(body);
+	}
+	{
+
+		btBoxShape* colShape = new btBoxShape(btVector3(.1, .1, .1));
+
+		btTransform startTransform;
+		startTransform.setIdentity();
+
+		btScalar mass(1.f);
+
+		bool isDynamic = (mass != 0.f);
+
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			colShape->calculateLocalInertia(mass, localInertia);
+		for (int i = 0; i < 3; i++)
+		{
+			int a[4];
+			if (i == 0)
+			{
+				a[0] = 2;
+				a[1] = 0;
+				a[2] = -1;
+				a[3] = 1;
+			}
+			else if (i ==1)
+			{
+				a[0] = 2;
+				a[1] = 2;
+				a[2] = -1;
+				a[3] = -1;
+			}
+			else
+			{
+				a[0] = 0;
+				a[1] = 1;
+				a[2] = 1;
+				a[3] = 0;
+			}
+			startTransform.setOrigin(btVector3(
+				btScalar(a[0]),
+				btScalar(10),
+				btScalar(a[1])));
+			#ifdef USE_MOTIONSTATE
+				btDefaultMotionState* myMotionState = new btDefaultMotionState(startTransform);
+
+				btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, colShape, localInertia);
+
+				box_before[i] = new btRigidBody(cInfo);
+			#else
+				box_before[i] = new btRigidBody(mass, 0, colShape, localInertia);
+				box_before[i]->setWorldTransform(startTransform);
+			#endif
+			box_before[i]->setUserIndex(-1);
+			initial_world->addRigidBody(box_before[i]);
+			box_before[i]->setLinearVelocity(btVector3(a[2], 0, a[3]));
+		}
+	}
+
+	// step simulation until collision
+
+	const double delta_t = 0.01;
+	int time = 0;
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
+
+	// serialize
+
+	printf("Serialize in %d time\n", time);
+	btSerializer* s = new btDefaultSerializer;
+	initial_world->serialize(s);
+	FILE* file = fopen(filename, "wb");
+	fwrite(s->getBufferPointer(), s->getCurrentBufferSize(), 1, file);
+	fclose(file);
+	const btVector3 before_serialize1 = box_before[0]->getCenterOfMassPosition();
+	const btVector3 before_serialize2 = box_before[1]->getCenterOfMassPosition();
+	const btVector3 before_serialize3 = box_before[2]->getCenterOfMassPosition();
```

Comment:
Good that you made these const. Let's move these 3 lines to just before the serialize, because we want to be confident that nothing in the serialization pipeline changes the position of the objects.